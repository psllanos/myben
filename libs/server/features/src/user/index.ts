export * from './user.processor'
export * from './user.service'
export * from './onboarding.schema'

export type { OnboardingState, RegisteredStep } from './onboarding.service'
