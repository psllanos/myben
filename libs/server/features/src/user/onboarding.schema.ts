import { z } from 'zod'

export const UpdateOnboardingSchema = z.discriminatedUnion('flow', [
    z.object({
        flow: z.literal('main'),
        updates: z
            .object({
                key: z.enum([
                    'intro',
                    'profile',
                    'verifyEmail',
                    'firstAccount',
                    'accountSelection',
                    'terms',
                    'maybe',
                    'welcome',
                ]),
                markedComplete: z.boolean(),
            })
            .array(),
    }),
    z.object({
        flow: z.literal('sidebar'),
        markedComplete: z.boolean().optional(),
        updates: z
            .object({
                key: z.enum([
                    'connect-depository',
                    'connect-investment',
                    'connect-liability',
                    'add-crypto',
                    'add-property',
                    'add-vehicle',
                    'add-other',
                    'upgrade-account',
                    'create-plan',
                ]),
                markedComplete: z.boolean(),
            })
            .array(),
    }),
])
