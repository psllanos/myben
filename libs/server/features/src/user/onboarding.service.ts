import type { SharedType } from '@maybe-finance/shared'

export type RegisteredStep = {
    key: string
    markedComplete: boolean
}

// DB Json field where each key represents a separate onboarding flow
export type OnboardingState = {
    [key in SharedType.OnboardingFlow]: {
        markedComplete: boolean // An override to indicate the flow is complete
        steps: RegisteredStep[]
    }
}

type CallbackFn<TData = unknown, T = boolean> = (user: TData, step: Step<TData>) => T

export class Step<TData = unknown> {
    group: string | undefined
    ctaPath: string | undefined
    title!: CallbackFn<TData, string>
    isComplete!: CallbackFn<TData>
    isMarkedComplete: CallbackFn<TData> = () => false
    isExcluded: CallbackFn<TData> = () => false
    isOptional = false

    /**
     * Constructs an onboarding step
     * @param key a stable key that the UI relies to show relevant view components
     * @param group a grouping that UI uses to visually group steps as "substeps"
     */
    constructor(readonly key: string) {}

    addToGroup(group: string) {
        this.group = group
        return this
    }

    /** When clicked, the router path that app will redirect to */
    setCTAPath(path: string) {
        this.ctaPath = path
        return this
    }

    setTitle(fn: CallbackFn<TData, string>) {
        this.title = fn
        return this
    }

    completeIf(fn: CallbackFn<TData>) {
        this.isComplete = fn
        return this
    }

    markedCompleteIf(fn: CallbackFn<TData>) {
        this.isMarkedComplete = fn
        return this
    }

    excludeIf(fn: CallbackFn<TData>) {
        this.isExcluded = fn
        return this
    }

    optional() {
        this.isOptional = true
        return this
    }
}

export class Onboarding<TData = unknown> {
    private _steps: Step<TData>[] = []

    constructor(
        private readonly data: TData,
        private readonly onboardingCompleteOverride: boolean
    ) {
        if (!this._steps.every((step) => step.completeIf != null && step.title != null)) {
            throw new Error('Every step must define completeIf callback fn and title')
        }
    }

    get isComplete() {
        return this.steps.every(
            (step) => step.isComplete || step.isOptional || step.isMarkedComplete
        )
    }

    get isMarkedComplete() {
        return this.steps.every((step) => step.isMarkedComplete) || this.onboardingCompleteOverride
    }

    /** Progress, expressed as percentage. */
    get progress() {
        const completed = this.steps.filter(
            (step) => step.isComplete || step.isMarkedComplete
        ).length

        return {
            completed,
            total: this.steps.length,
            percent: +(completed / this.steps.length).toFixed(2),
        }
    }

    get steps() {
        return this._steps
            .filter((step) => !step.isExcluded(this.data, step))
            .map((step) => ({
                key: step.key,
                title: step.title(this.data, step),
                group: step.group,
                ctaPath: step.ctaPath,
                isOptional: step.isOptional,
                isComplete: step.isComplete(this.data, step),
                isMarkedComplete: step.isMarkedComplete(this.data, step),
            }))
    }

    get currentStep() {
        const incompleteSteps = this.steps.filter((step) => !step.isComplete)
        return incompleteSteps[0]
    }

    addStep(key: string) {
        const step = new Step<TData>(key)
        this._steps.push(step)
        return step
    }
}
