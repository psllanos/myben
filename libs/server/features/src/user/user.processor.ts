import type { Logger } from 'winston'
import type { PrismaClient, User } from '@prisma/client'
import type { SyncUserQueueJobData } from '@maybe-finance/server/shared'
import type { IAccountService } from '../account'
import type { IUserService } from './user.service'
import type {
    IAccountConnectionProviderFactory,
    IAccountConnectionService,
} from '../account-connection'
import { ServerUtil } from '@maybe-finance/server/shared'

export interface IUserProcessor {
    sync(jobData: SyncUserQueueJobData): Promise<void>
    delete(jobData: SyncUserQueueJobData): Promise<void>
}

export class UserProcessor implements IUserProcessor {
    constructor(
        private readonly logger: Logger,
        private readonly prisma: PrismaClient,
        private readonly userService: IUserService,
        private readonly accountService: IAccountService,
        private readonly connectionService: IAccountConnectionService,
        private readonly connectionProviders: IAccountConnectionProviderFactory
    ) {}

    async sync(jobData: SyncUserQueueJobData) {
        const user = await this.userService.get(jobData.userId)

        await ServerUtil.useSync<User>({
            sync: async (user) => {
                const { connections, accounts } = await this.accountService.getAll(user.id)

                await Promise.allSettled([
                    ...connections.map((connection) => this.connectionService.sync(connection.id)),
                    ...accounts.map((account) => this.accountService.sync(account.id)),
                ])
            },
            onSyncSuccess: (user) => this.userService.syncBalances(user.id),
            onSyncError: async (user, error) => {
                this.logger.error(`error syncing user ${user.id}`, { error })
            },
        })(user)
    }

    async delete(jobData: SyncUserQueueJobData) {
        const { userId } = jobData

        this.logger.info(`deleting user ${userId}...`)

        const connections = await this.prisma.accountConnection.findMany({
            where: { userId },
        })

        // delete connection data
        this.logger.info(`deleting user ${userId} data for ${connections.length} connections...`)
        await Promise.allSettled(connections.map((c) => this.connectionProviders.for(c).delete(c)))

        // delete from database (will cascade to relations)
        this.logger.info(`deleting user ${userId} from database...`)
        await this.prisma.user.delete({ where: { id: userId } })

        this.logger.info(`user ${userId} deleted successfully`)
    }
}
