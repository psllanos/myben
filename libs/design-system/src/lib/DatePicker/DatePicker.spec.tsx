import user from '@testing-library/user-event'
import { fireEvent, render, screen, waitFor } from '@testing-library/react'
import { DatePicker } from './'
import { DateTime } from 'luxon'

// DatePicker configuration
const minDate = DateTime.now().minus({ years: 2 })
const maxDate = DateTime.now()

describe('<DatePicker />', () => {
    describe('When datepicker is closed', () => {
        it('should have placeholder when empty', () => {
            const onChangeMock = jest.fn()

            const component = render(
                <DatePicker
                    name="picker"
                    minCalendarDate={minDate.toISODate()}
                    maxCalendarDate={maxDate.toISODate()}
                    value={null}
                    onChange={onChangeMock}
                />
            )

            expect(screen.getByPlaceholderText('MM / DD / YYYY')).toBeInTheDocument()
            expect(screen.queryByText('Cancel')).not.toBeInTheDocument()
            expect(screen.queryByText('Apply')).not.toBeInTheDocument()
            expect(component).toMatchSnapshot()
        })

        it('should have date when valid', () => {
            const onChangeMock = jest.fn()

            const component = render(
                <DatePicker
                    name="picker"
                    minCalendarDate={minDate.toISODate()}
                    maxCalendarDate={maxDate.toISODate()}
                    value={null}
                    onChange={onChangeMock}
                />
            )

            const input = component.getByPlaceholderText('MM / DD / YYYY') as HTMLInputElement

            user.type(input, '02012021')
            expect(onChangeMock).toHaveBeenCalledWith('2021-02-01')
            expect(screen.queryByText('Date must be')).not.toBeInTheDocument()

            expect(screen.queryByText('Cancel')).not.toBeInTheDocument()
            expect(screen.queryByText('Apply')).not.toBeInTheDocument()
        })
    })

    describe('When datepicker is expanded', () => {
        it('should be able to open modal and select a date', async () => {
            const onChangeMock = jest.fn()

            render(
                <div id="container">
                    <DatePicker
                        name="picker"
                        minCalendarDate={minDate.toISODate()}
                        maxCalendarDate={maxDate.toISODate()}
                        value={null}
                        onChange={onChangeMock}
                    />
                </div>
            )

            const currentMonth = DateTime.now()
            const priorMonth = DateTime.now().minus({ months: 1 })

            // Open the modal
            fireEvent.click(screen.getByTestId('datepicker-toggle-icon'))

            await waitFor(() => expect(screen.getByTestId('datepicker-panel')).toBeInTheDocument())

            // Go back a month
            fireEvent.click(screen.getByTestId('datepicker-range-back-arrow'))
            expect(screen.queryByText(priorMonth.monthShort)).toBeInTheDocument()
            expect(screen.queryByText(priorMonth.year)).toBeInTheDocument()

            // Go forward a month
            fireEvent.click(screen.getByTestId('datepicker-range-next-arrow'))
            expect(screen.queryByText(currentMonth.monthShort)).toBeInTheDocument()
            expect(screen.queryByText(currentMonth.year)).toBeInTheDocument()

            // Select a date
            fireEvent.click(screen.getByText('17'))
            fireEvent.click(screen.getByText('Apply'))
            expect(onChangeMock).toHaveBeenCalledWith(
                DateTime.fromObject(
                    {
                        day: 17,
                        month: currentMonth.month,
                        year: currentMonth.year,
                    },
                    { zone: 'utc' }
                ).toISODate()
            )
        })
    })
})
