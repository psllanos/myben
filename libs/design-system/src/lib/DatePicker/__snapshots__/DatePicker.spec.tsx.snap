// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<DatePicker /> When datepicker is closed should have placeholder when empty 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="relative"
        data-headlessui-state=""
      >
        <div>
          <label
            class="flex w-full flex-col"
          >
            <div
              class="flex h-10 text-white rounded border overflow-hidden relative border-gray-700 focus-within:ring focus-within:ring-opacity-10 focus-within:border-cyan focus-within:ring-cyan"
            >
              <input
                class="min-w-0 py-0 w-full text-base font-light leading-none border-0 focus:outline-none focus:ring-0 placeholder-gray-100 disabled:placeholder-gray-200 disabled:text-gray-100 bg-gray-500 pl-3 pr-8"
                inputmode="numeric"
                name="picker"
                placeholder="MM / DD / YYYY"
                type="text"
                value=""
              />
              <span
                class="absolute right-3 top-1/2 -translate-y-1/2 flex items-center justify-center text-gray-50 text-base select-none bg-gray-500"
              >
                <button
                  aria-expanded="false"
                  data-headlessui-state=""
                  data-testid="datepicker-toggle-icon"
                  id="headlessui-popover-button-:r0:"
                  type="button"
                >
                  <svg
                    class="text-lg"
                    fill="currentColor"
                    height="1em"
                    stroke="currentColor"
                    stroke-width="0"
                    viewBox="0 0 24 24"
                    width="1em"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <g>
                      <path
                        d="M0 0h24v24H0z"
                        fill="none"
                      />
                      <path
                        d="M17 3h4a1 1 0 0 1 1 1v16a1 1 0 0 1-1 1H3a1 1 0 0 1-1-1V4a1 1 0 0 1 1-1h4V1h2v2h6V1h2v2zM4 9v10h16V9H4zm2 4h5v4H6v-4z"
                      />
                    </g>
                  </svg>
                </button>
              </span>
            </div>
          </label>
        </div>
      </div>
    </div>
    <div
      id="headlessui-portal-root"
    >
      <div
        data-headlessui-portal=""
      />
    </div>
  </body>,
  "container": <div>
    <div
      class="relative"
      data-headlessui-state=""
    >
      <div>
        <label
          class="flex w-full flex-col"
        >
          <div
            class="flex h-10 text-white rounded border overflow-hidden relative border-gray-700 focus-within:ring focus-within:ring-opacity-10 focus-within:border-cyan focus-within:ring-cyan"
          >
            <input
              class="min-w-0 py-0 w-full text-base font-light leading-none border-0 focus:outline-none focus:ring-0 placeholder-gray-100 disabled:placeholder-gray-200 disabled:text-gray-100 bg-gray-500 pl-3 pr-8"
              inputmode="numeric"
              name="picker"
              placeholder="MM / DD / YYYY"
              type="text"
              value=""
            />
            <span
              class="absolute right-3 top-1/2 -translate-y-1/2 flex items-center justify-center text-gray-50 text-base select-none bg-gray-500"
            >
              <button
                aria-expanded="false"
                data-headlessui-state=""
                data-testid="datepicker-toggle-icon"
                id="headlessui-popover-button-:r0:"
                type="button"
              >
                <svg
                  class="text-lg"
                  fill="currentColor"
                  height="1em"
                  stroke="currentColor"
                  stroke-width="0"
                  viewBox="0 0 24 24"
                  width="1em"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <g>
                    <path
                      d="M0 0h24v24H0z"
                      fill="none"
                    />
                    <path
                      d="M17 3h4a1 1 0 0 1 1 1v16a1 1 0 0 1-1 1H3a1 1 0 0 1-1-1V4a1 1 0 0 1 1-1h4V1h2v2h6V1h2v2zM4 9v10h16V9H4zm2 4h5v4H6v-4z"
                    />
                  </g>
                </svg>
              </button>
            </span>
          </div>
        </label>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
